### ISSUE TITLE
Opportunistic Session Termination

### ISSUE THREAD

**Issue by tghosth** [OWASP/ASVS#1800]
In #1417 and #1641, we added a new requirement [3.3.6](https://github.com/OWASP/ASVS/blob/master/5.0/en/0x12-V3-Session-management.md#v33-session-termination) related to session termination:

| # | Description | L1 | L2 | L3 | CWE | [NIST &sect;](https://pages.nist.gov/800-63-3/sp800-63b.html) |
| :---: | :--- | :---: | :---: | :---: | :---: | :---: |
| **3.3.6** | [ADDED] Verify that all active sessions are revoked when a user account is disabled or deleted (such as an employee leaving the company). | ✓ | ✓ | ✓ | 613 | |

This joins some existing requirements which also deal with session termination but appear to be focused on either automated functionality after credentials are changed or functionality available to the application user.

| # | Description | L1 | L2 | L3 | CWE | [NIST &sect;](https://pages.nist.gov/800-63-3/sp800-63b.html) |
| :---: | :--- | :---: | :---: | :---: | :---: | :---: |
| **3.3.3** | [MODIFIED, LEVEL L2 > L1] Verify that the application gives the option to terminate all other active sessions after a successful change or removal of any authentication factor (including password change via reset or recovery and, if present, an MFA settings update). This must be effective across the application, federated login (if present), and any relying parties. | ✓ | ✓ | ✓ | 613 | |
| **3.3.4** | Verify that users are able to view and (having re-entered login credentials) log out of any or all currently active sessions and devices. | | ✓ | ✓ | 613 | 7.1 |

Should we create a new requirement to make it clear that an application administrator should also be able to terminate all or some of a user's sessions, if necessary?

**Comment by elarlang** [comment]
> I would ask the same question why do we have idle timeout, then Elar? Closing sessions, when an account is locked is a similar problem, it’s a clear necessary requirement for secure software.

It makes sense to ask it to address the requirement better and to recheck, do we have something already in the requirement list to cover that? It does not mean anyhow, that it is not necessary.

**Comment by elarlang** [comment]
I am just documenting one more idea: if this proposal will be added as a requirement, then maybe it makes sense to create a new section for V3 "Session Administration" or something like that.

Additionally to this issue, we can move the added requirement:
* _V3.3.6 [ADDED] Verify that all active sessions are revoked when a user account is disabled or deleted (such as an employee leaving the company)._

**Comment by elarlang** [comment]
I think there is an agreement here to move on with that, can you propose the requirement text @tghosth ?

**Comment by tghosth** [comment]
I can support creating a dedicated section.

To be clear, the reason for the requirement is that it may be useful to be able to force session revocation, e.g if tokens are stolen or, without actually disabling the user.

Also, there is an edge case where the user's active sessions stop working when the user is disabled but do they start working again after the user is re-enabled.

Suggested wording:

**_[ADDED] Verify that the application makes it possible to revoke some or all active sessions to force re-authentication._**

**Comment by tghosth** [comment]
Handle a situation where some sort of SSO service handles session management:

**_[ADDED] Verify that the application's session management functionality makes it possible to revoke some or all active sessions to force re-authentication._**

**Comment by jmanico** [comment]
+1 to this direction

**Comment by elarlang** [comment]
discussion over PR...

Commit related to this issue: https://github.com/OWASP/ASVS/pull/1819/commits/7913e03fc68dc22000a082bba304a5e2fc1a67ed

This wording is a bit messy:
_Verify that application administrators are able to terminate any or all currently active sessions and devices for a user or for all users._

Why not just simple?
_Verify that application administrators are able to terminate sessions._

**Comment by tghosth** [comment]
So I wanted to be specific about the expected functionality. I think I could settle for:

_Verify that application administrators are able to terminate active sessions for an individual user or for all users._

**Comment by elarlang** [comment]
Is there any reason to move from "the application provides functionality" to "application administrators are able to"?

An admin can go directly to the database and delete it with an SQL query, but this is not what we want to achieve.

**Comment by elarlang** [comment]
Maybe should have separate issues for the sections? At the moment there are 2 parallel topics.

I don't mind splitting requirements into different sections like it is proposed via PR at the moment, but there are some other ways to keep in mind (level: nice to have):
* single-session application vs multi-session application - many requirements are valid only for multi-session applications. It would be nice if those are together and are easy to skip if the application allows only one session per user. Valid use-case for example when authentication requires "something you have" such as ID-card or yubikey, in this case, you can not have more than one parallel session.
* the application user vs the application administrator - often for pen-tests we don't have access to administrator systems and we can test only the application functionality. again, it would be nice, if those are separately and are easy to skip.

Should we ask clearly with a separate requirement, that the application must have a clear strategy, how many sessions it allows per single user?

**Comment by jmanico** [comment]
> Also, there is an edge case where the user's active sessions stop working when the user is disabled but do they start working again after the user is re-enabled.

Josh I politely disagree. Once a session is disabled, I think it should be permanent and the user would need to re-authenticate.

**Comment by jmanico** [comment]
An Josh, I like the more extensive requirement and suggest:

_Verify that application administrators can terminate any active user sessions and connected devices, for individual users or all users._

**Comment by elarlang** [comment]
I wrote it to PR comment, but I repeat it here - if we mention "connected devices", it is a hidden (not clearly written) requirement, that the application need to track devices. I don't think it should be the case for L1 and maybe L2 applications.

**Comment by jmanico** [comment]
> I wrote it to PR comment, but I repeat it here - if we mention "connected devices", it is a hidden (not clearly written) requirement, that the application need to track devices. I don't think it should be the case for L1 and maybe L2 applications.

L2 works for me, that makes sense.``

**Comment by elarlang** [comment]
There are many parallel topics here, it is a bit messy and it is quite hard to follow. This is an example of why we have agreed "agreement in the issue first, PR later" process and why we should follow it.

I prefer we have this issue for the new requirement and a new separate issue for section re-org and maybe a new separate issue for language alignment or changes to existing requirements. It's a bit messy at the moment and it will be complicated to follow the reasoning for the changes later.

**Comment by tghosth** [comment]
> > I wrote it to PR comment, but I repeat it here - if we mention "connected devices", it is a hidden (not clearly written) requirement, that the application need to track devices. I don't think it should be the case for L1 and maybe L2 applications.
> 
> L2 works for me, that makes sense.``

Opened #1829 to discuss this.

**Comment by tghosth** [comment]
@elarlang to review PR #1819 again to see if there are further comments
